File: src/chasm/primitives/transaction.py
class Transaction:
    class Output:
        def __init__(self):
            self.value: int

    class XpeerOutput:
        def __init__(self):
            self.offer_hash: bytes
            self.sender: bytes
            self.receiver: bytes
            self.timestamp: int

    class TransferOutput:
        def __init__(self):
            self.receiver: bytes

    class Input:
        def __init__(self):
            self.tx_hash: bytes
            self.index: int
            self.signature: bytes[64]
            self.utxo: Transaction.Output

    def __init__(self):
        self.inputs: [Transaction.Input] = []
        self.outputs: [Transaction.Output] = []
        self.bytes: bytes = bytes

File: src/chasm/primitives/Transaction.cpp
//
// Created by Daniel Bigos on 11.11.18.
//

#include <chasm/serialization/Serializer.hpp>
#include "Transaction.hpp"

using namespace chasm;
using namespace chasm::primitives;

void Transaction::addInput(hash_t const &txHash, in_idx_t input) {
    if (inputs_.size() > std::numeric_limits<x_size_t>::max()) throw std::length_error("Too many inputs");
    inputs_.push_back(std::make_unique<transaction::Input>(txHash, input));
}

void Transaction::addInput(hash_t &&txHash, in_idx_t input) {
    if (inputs_.size() > std::numeric_limits<x_size_t>::max()) throw std::length_error("Too many inputs");
    inputs_.push_back(std::make_unique<transaction::Input>(txHash, input));
}

void Transaction::addOutput(value_t value, address_t const &receiver) {
    if (outputs_.size() > std::numeric_limits<x_size_t>::max()) throw std::length_error("Too many outputs");
    outputs_.push_back(std::make_unique<transaction::SimpleOutput>(value, receiver));;
}



const transaction::Input &Transaction::getInput(x_size_t index) const {
    return *inputs_.at(index);
}

const transaction::Output &Transaction::getOutput(x_size_t index) const {
    return *outputs_.at(index);
}

//void Transaction::accept(serialization::Serializer &serializer) {
//    serializer.visit(*this);
//}
